File: python/examples/02_cite_check.py
"""Example 02 -- Cite check citations (Python).

Match expanded rows on inputIndex (not array index). There is no cite
verdict "not_found" -- use not_in_corpus / implausible / not_covered / etc.
"""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from lawdiver import LawDiverClient


def summarize_negatives(item: dict) -> str | None:
    verdict = item.get("verdict")
    if verdict == "implausible":
        return "strong fabrication signal"
    if verdict == "not_in_corpus":
        return "searched held range, no match"
    if verdict == "not_covered":
        return "range not held / unparseable"
    if verdict == "unverified":
        return "cannot confirm or deny"
    if verdict == "error":
        if item.get("lookupStatus") == "deadline_exceeded":
            return "soft timeout (deadline_exceeded) -- not billed; retry"
        return "row failed -- not billed"
    return None


with LawDiverClient() as client:
    payload = client.cite_check(citations=["570 U.S. 744", "999 F.3d 1"])

print(f"requestId: {payload['requestId']}")
print(f"units: {payload['usage']['quantity']}")
print(f"rows: {len(payload['results'])} (may exceed input count for compounds)")
print()

for item in payload["results"]:
    label = item.get("citationAsSent") or item.get("citationAsWritten") or "(unknown)"
    idx = item.get("inputIndex", "?")
    unit = item.get("unitIndex")
    unit_bit = f" unit {unit}" if unit is not None else ""
    print(f"[input {idx}{unit_bit}] {label} -> {item['verdict']}")
    if item.get("lookupStatus"):
        print(f"  lookupStatus: {item['lookupStatus']}")
    if item.get("correctedCitation"):
        print(f"  Bluebook: {item['correctedCitation']}")
    if item.get("explanation"):
        print(f"  {item['explanation']}")
    if item.get("corpusCaveat"):
        print(f"  caveat: {item['corpusCaveat']}")
    note = summarize_negatives(item)
    if note:
        print(f"  note: {note}")
    if item["verdict"] in ("likely_valid", "name_mismatch", "page_mismatch"):
        for cand in item.get("candidates") or []:
            print(f"  candidate: {cand.get('bluebookCitation') or cand}")
    print()

File: typescript/examples/02-cite-check.ts
/**
 * Example 02 -- Cite check one or many citations
 * Language: TypeScript (Node.js)
 * Docs: https://lawdiver.com/docs/api
 *
 * Match expanded rows on inputIndex (not array index). There is no cite
 * verdict "not_found" -- use not_in_corpus / implausible / not_covered / etc.
 */
import { LawDiverClient } from "../src/client.js";
import type { CiteCheckItem } from "../src/types.js";

const client = new LawDiverClient();

const payload = await client.citeCheck({
  citations: ["570 U.S. 744", "999 F.3d 1"],
});

console.log(`requestId: ${payload.requestId}`);
console.log(`units: ${payload.usage.quantity}`);
console.log(`rows: ${payload.results.length} (may exceed input count for compounds)`);
console.log("");

function summarizeNegatives(item: CiteCheckItem): string | null {
  switch (item.verdict) {
    case "implausible":
      return "strong fabrication signal";
    case "not_in_corpus":
      return "searched held range, no match";
    case "not_covered":
      return "range not held / unparseable";
    case "unverified":
      return "cannot confirm or deny";
    case "error":
      return item.lookupStatus === "deadline_exceeded"
        ? "soft timeout (deadline_exceeded) -- not billed; retry"
        : "row failed -- not billed";
    default:
      return null;
  }
}

for (const item of payload.results) {
  const label = item.citationAsSent ?? item.citationAsWritten ?? "(unknown)";
  console.log(
    `[input ${item.inputIndex}${item.unitIndex != null ? ` unit ${item.unitIndex}` : ""}] ${label} -> ${item.verdict}`,
  );
  if (item.lookupStatus) console.log(`  lookupStatus: ${item.lookupStatus}`);
  if (item.correctedCitation) console.log(`  Bluebook: ${item.correctedCitation}`);
  if (item.explanation) console.log(`  ${item.explanation}`);
  if (item.corpusCaveat) console.log(`  caveat: ${item.corpusCaveat}`);

  const negative = summarizeNegatives(item);
  if (negative) console.log(`  note: ${negative}`);

  if (
    item.verdict === "likely_valid" ||
    item.verdict === "name_mismatch" ||
    item.verdict === "page_mismatch"
  ) {
    for (const c of item.candidates ?? []) {
      console.log(`  candidate: ${c.bluebookCitation ?? JSON.stringify(c)}`);
    }
  }
  console.log("");
}
